eth/validation.py

```diff
@@ -190,14 +190,21 @@ def validate_uint256(value: int, title: str="Value") -> None:
         )
 
 
-def validate_stack_item(value: Union[int, bytes]) -> None:
-    if isinstance(value, bytes) and len(value) <= 32:
+def validate_stack_int(value: int) -> None:
+    if 0 <= value <= UINT_256_MAX:
         return
-    elif isinstance(value, int) and 0 <= value <= UINT_256_MAX:
+    raise ValidationError(
+        "Invalid Stack Item: Must be either a length 32 byte "
+        "string or a 256 bit integer. Got {!r}".format(value)
+    )
+
+
+def validate_stack_bytes(value: bytes) -> None:
```

Comment:
I'd advocate for going ahead and dropping support for heterogenous stack and update so everything uses an `int` but that doesn't have to happen here, but I suspect there's more easy performance to be gained there.

Reply:
Hah, so that's a tangent I went on and ended up dropping. I got some early indications that the extra cost of converting and de-converting the elements when both the push and pop are the same type (say, both `bytes`) is meaningful.

Reply:
Maybe you could code comment that in the main class so that it's documented that a heterogeneous stack is more performant.